testing/dart_isolate_runner.h

```diff
@@ -0,0 +1,198 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/make_copyable.h"
+#include "flutter/fml/paths.h"
+#include "flutter/fml/synchronization/waitable_event.h"
+#include "flutter/fml/thread.h"
+
+namespace flutter {
+namespace testing {
+
+class AutoIsolateShutdown {
+ public:
+  AutoIsolateShutdown() = default;
+
+  AutoIsolateShutdown(std::shared_ptr<DartIsolate> isolate,
+                      fml::RefPtr<fml::TaskRunner> runner)
+      : isolate_(std::move(isolate)), runner_(std::move(runner)) {}
+
+  ~AutoIsolateShutdown() {
+    if (!IsValid()) {
+      return;
+    }
+    fml::AutoResetWaitableEvent latch;
+    fml::TaskRunner::RunNowOrPostTask(
+        runner_, [isolate = std::move(isolate_), &latch]() {
+          if (!isolate->Shutdown()) {
+            FML_LOG(ERROR) << "Could not shutdown isolate.";
+            FML_CHECK(false);
+          }
+          latch.Signal();
+        });
+    latch.Wait();
+  }
+
+  bool IsValid() const { return isolate_ != nullptr && runner_; }
+
+  FML_WARN_UNUSED_RESULT
+  bool RunInIsolateScope(std::function<bool(void)> closure) {
+    if (!IsValid()) {
+      return false;
+    }
+
+    bool result = false;
+    fml::AutoResetWaitableEvent latch;
+    fml::TaskRunner::RunNowOrPostTask(
+        runner_, [this, &result, &latch, closure]() {
+          tonic::DartIsolateScope scope(isolate_->isolate());
+          tonic::DartApiScope api_scope;
+          if (closure) {
+            result = closure();
+          }
+          latch.Signal();
+        });
+    latch.Wait();
+    return true;
+  }
+
+  DartIsolate* get() {
+    FML_CHECK(isolate_);
+    return isolate_.get();
+  }
+
+ private:
+  std::shared_ptr<DartIsolate> isolate_;
+  fml::RefPtr<fml::TaskRunner> runner_;
+
+  FML_DISALLOW_COPY_AND_ASSIGN(AutoIsolateShutdown);
+};
+
+static void RunDartCodeInIsolate(DartVMRef& vm_ref,
```

Comment:
As soon as this header gets included in more than one translation unit, there are going to be duplicate symbol errors. The implementation needs to be in its own TU.